### PR TITLE
feat: allow use of empty alg fields in jwks

### PR DESCRIPTION
**Pull request by SkArchon**
**Current Behaviour**
Currently we require "ALG" to be present in the JWKS, however this is optional as per the rfc. We filter out a default list of algorithms / passed algorithms, based on the alg parameter, but if the alg parameter is "", this is skipped, which leads us to this issue.

**What this PR does**
This PR cleans up the validation store and we validate the algorithm on the hot path now instead if an algorithm HAS been specified. In order to prevent repeating logic of getting and casting the algorithm, we looked at forking the keyfunc from https://github.com/MicahParks/keyfunc (this is simple since its one file we need to maintain and it gets rarely updated, it is also the expectation of the author for users to fork the keyfunc for customization, [reference](https://github.com/MicahParks/jwkset/issues/54#issuecomment-3108240778)). However as its more likely the router will get security reports, we opted not to do this.

This PR also upgrades the jwkset library to include this [downstream change](https://github.com/MicahParks/jwkset/pull/61/files)

To summarise, this PR will not filter JWKS on startup / refresh, this could increase potential memory usage by a few kb, we removed the validation store for simplicity. And moves the validation of the algorithm to the hot path, this adds a few microseconds to every request. The validation for loop can run for a max of 12 cycles since we only allow 12 algorithms.

~~Prior PR: https://github.com/wundergraph/keyfunc/pull/1~~

## Summary by CodeRabbit

* **New Features**
  * Per-provider audience and algorithm enforcement plus a new configurable refresh_unknown_kid option for JWKS sources.

* **Tests**
  * Greatly expanded authentication tests: multi-provider audience cases, algorithm families and overrides, unknown-KID refresh flows, concurrent requests, malformed/edge-case tokens.

* **Chores**
  * Upgraded JWKS/JWT dependencies for improved compatibility.

* **Breaking/Removal**
  * Removed legacy JWKS validation wrapper; validation now uses the new per-entry approach.

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SkArchon** [comment]
@coderabbitai review

**Comment by SkArchon** [comment]
@coderabbitai review

**Comment by SkArchon** [comment]
@coderabbitai review

File: router/pkg/authentication/jwks_token_decoder.go [hunk 1]
```diff
@@ -151,14 +164,40 @@
...
+					continue
+				}
+
+				// This is a custom validation different from keyfunc.Keyfunc
+				if !slices.Contains(keyFuncAndOpts.allowedAlgorithms, alg) {
+					errJoin = errors.Join(errJoin, fmt.Errorf("%w: could not find alg %s in allow list", keyfunc.ErrKeyfunc, alg))
+					continue
+				}
+			}
+
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/jwks_token_decoder.go]: This is fine since we return `store.getSupportedAlgorithms()` from the validation_store

File: router/pkg/authentication/jwks_token_decoder.go [hunk 2]
```diff
@@ -46,7 +47,8 @@ type JWKSConfig struct {
 	Algorithm string
 	KeyId     string
 
-	Audiences []string
+	Audiences           []string
+	AllowEmptyAlgorithm bool
 }
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/jwks_token_decoder.go]: This is already exposed

File: router-tests/jwks/crypto.go [hunk 3]
```diff
@@ -43,11 +43,14 @@ func (b *baseCrypto) MarshalJWK() (jwkset.JWK, error) {
 	}
 
 	meta := jwkset.JWKMetadataOptions{
-		ALG: b.alg,
 		KID: b.kID,
 		USE: jwkset.UseSig,
 	}
 
+	if b.alg != "" {
```
Comment (SkArchon) [inline router-tests/jwks/crypto.go]: can remove since its anyways going to use the default value.

**Comment by StarpTech** [comment]
If the algorithm is not required according to the spec, we should follow the spec and only validate it when explicitly set.

**Comment by SkArchon** [comment]
@coderabbitai review

File: router/pkg/authentication/keyfunc/keyfunc_test.go [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+	// Parse the JWT.
+	token, err := jwt.Parse(jwtB64, k.Keyfunc)
+	if err != nil {
+		t.Fatalf("Failed to parse the JWT.\nError: %s", err)
+	}
+
+	// Check if the token is valid.
+	if !token.Valid {
+		t.Fatalf("The token is not valid.")
+	}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/keyfunc/keyfunc_test.go]: this is from forked code, ignore

File: router/pkg/authentication/keyfunc/keyfunc_test.go [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/keyfunc/keyfunc_test.go]: this is from forked code, ignore

File: router-tests/jwks/jwks.go [hunk 6]
```diff
@@ -36,13 +36,31 @@ func (s *Server) Close() {
...
+		if tokenOpts.AlgOverride != "" {
+			method := jwt.GetSigningMethod(tokenOpts.AlgOverride)
+			if method == nil {
+				return "", fmt.Errorf("unsupported signing method: %s", tokenOpts.AlgOverride)
+			}
+			token = jwt.NewWithClaims(method, jwt.MapClaims(claims))
+		} else {
+			token = jwt.NewWithClaims(pr.SigningMethod(), jwt.MapClaims(claims))
+		}
+
```
Comment (Noroth) [inline router-tests/jwks/jwks.go]: Minor simplification suggestion to remove the else block

```suggestion
		var token *jwt.Token
		method := pr.SigningMethod()
		if tokenOpts.AlgOverride != "" {
			method = jwt.GetSigningMethod(tokenOpts.AlgOverride)
			if method == nil {
				return "", fmt.Errorf("unsupported signing method: %s", tokenOpts.AlgOverride)
			}
		}
		
		token = jwt.NewWithClaims(method, jwt.MapClaims(claims))
```

File: router/pkg/authentication/keyfunc/keyfunc.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
+// This is forked from https://github.com/MicahParks/keyfunc/blob/main/keyfunc.go
+// Copyrights go to the original author.
+package keyfunc
```
Comment (Noroth) [inline router/pkg/authentication/keyfunc/keyfunc.go]: Wouldn't it make more sense to fork the repository and use a `replace` in go.mod? Then we could keep track of changes in the upstream repository more easily instead of trying to copy and paste the code and not override updated logic?

File: router/pkg/authentication/keyfunc/keyfunc.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+		algInter, ok := token.Header["alg"]
+		if !ok {
+			return nil, fmt.Errorf("%w: could not find alg in JWT header", ErrKeyfunc)
+		}
+		alg, ok := algInter.(string)
+		if !ok {
+			// For test coverage purposes, this should be impossible to reach because the JWT package rejects a token
+			// without an alg parameter in the header before calling jwt.Keyfunc.
+			return nil, fmt.Errorf(`%w: the JWT header did not contain the "alg" parameter, which is required by RFC 7515 section 4.1.1`, ErrKeyfunc)
+		}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/keyfunc/keyfunc.go]: this is from forked code, the test/config mismatch is resolved. Could you elaborate? The first point you mentioned is important since jwt should have alg, and if the router specifies an allowed alg it should match that for your third point.

File: router/pkg/authentication/jwks_token_decoder.go [hunk 9]
```diff
@@ -67,9 +68,14 @@ type audKey struct {
 
 type audienceSet map[string]struct{}
 
+type keyFuncWithOpts struct {
```
Comment (SkArchon) [inline router/pkg/authentication/jwks_token_decoder.go:71]: TODO: Cleanup

File: router/pkg/authentication/jwks_token_decoder.go [hunk 10]
```diff
@@ -164,18 +172,22 @@ func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKS
...
 		}
 	}
 
 	keyFuncWrapper := jwt.Keyfunc(func(token *jwt.Token) (any, error) {
 		var errJoin error
-		for key, keyFunc := range keyFuncMap {
-			pub, err := keyFunc.Keyfunc(token)
+		for key, keyFuncAndOpts := range keyFuncMap {
+
+			pub, err := keyFuncAndOpts.keyFunc.Keyfunc(token)
```
Comment (Noroth) [inline router/pkg/authentication/jwks_token_decoder.go]: Next refactor be like: `keyFuncAndOptsProvider.getKeyFunc().keyFunc.KeyFunc(token)`

Looks a bit funny. Maybe we can add a receiver function to the `keyFuncWithOpts` type that simply calls the `KeyFunc`?

File: router/pkg/authentication/keyfunc/keyfunc.go (hunk 7 above)
Comment (SkArchon) [reply router/pkg/authentication/keyfunc/keyfunc.go]: good idea

File: router-tests/jwks/jwks.go (hunk 6 above)
Comment (SkArchon) [reply router-tests/jwks/jwks.go]: Done

File: router/pkg/authentication/keyfunc/keyfunc.go (hunk 7 above)
Comment (SkArchon) [reply router/pkg/authentication/keyfunc/keyfunc.go]: Forked

File: router/pkg/authentication/jwks_token_decoder.go (hunk 10 above)
Comment (SkArchon) [reply router/pkg/authentication/jwks_token_decoder.go]: Done

**Comment by SkArchon** [comment]
Forked keyfunc, PR can be found here
PR: https://github.com/wundergraph/keyfunc/pull/1
cc: @StarpTech @Noroth

File: router/pkg/authentication/jwks_token_decoder.go [hunk 11]
```diff
@@ -185,10 +184,12 @@ func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKS
 			if len(expectedAudiences) > 0 {
 				tokenAudiences, err := token.Claims.GetAudience()
 				if err != nil {
-					return nil, fmt.Errorf("could not get audiences from token claims: %w", err)
+					errJoin = errors.Join(errJoin, fmt.Errorf("could not get audiences from token claims: %w", err))
```
Comment (StarpTech) [inline router/pkg/authentication/jwks_token_decoder.go:196]: Should we have a test for this?

**Comment by StarpTech** [comment]
https://github.com/wundergraph/keyfunc/pull/1#issuecomment-3321750302

**Comment by SkArchon** [comment]
cc: @StarpTech

File: router/pkg/authentication/jwks_token_decoder.go (hunk 11 above)
Comment (SkArchon) [reply router/pkg/authentication/jwks_token_decoder.go:196]: Done

**Review by StarpTech — APPROVED** [review]
LGTM

File: router/pkg/authentication/jwks_token_decoder.go [hunk 12]
```diff
@@ -168,29 +181,55 @@ func NewJwksTokenDecoder(ctx context.Context, logger *zap.Logger, configs []JWKS
...
+			if len(entry.allowedAlgorithms) > 0 {
+				algInter, ok := token.Header["alg"]
+				if !ok {
+					errJoin = errors.Join(errJoin, fmt.Errorf("%w: could not find alg in JWT header", keyfunc.ErrKeyfunc))
+					continue
+				}
+				alg, ok := algInter.(string)
+				if !ok {
+					errJoin = errors.Join(errJoin, fmt.Errorf(`%w: the JWT header did not contain the "alg" parameter, which is required by RFC 7515 section 4.1.1`, keyfunc.ErrKeyfunc))
+					continue
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/pkg/authentication/jwks_token_decoder.go:217]: this is the same string as the inner lib and cannot be changed.
